app/PaymentDrivers/Blockonomics/Blockonomics.php

```diff
@@ -137,43 +139,52 @@ public function paymentResponse(PaymentResponseRequest $request)
             'btc_address' => ['required'],
             'btc_amount' => ['required'],
             'btc_price' => ['required'],
-            // Setting status to required will break the payment process
-            // because sometimes the status is returned as 0 which is falsy
-            // and the validation will fail.
-            // 'status' => ['required'],
         ]);
 
+        $this->payment_hash = PaymentHash::where('hash', $request->payment_hash)->firstOrFail();
+
+        // Calculate fiat amount from Bitcoin
+        $amount_received_satoshis = $request->btc_amount;
+        $amount_satoshis_in_one_btc = 100000000;
+        $amount_received_btc = $amount_received_satoshis / $amount_satoshis_in_one_btc;
+        $price_per_btc_in_fiat = $request->btc_price;
+        $fiat_amount = round(($price_per_btc_in_fiat * $amount_received_btc), 2);
+
+        // Get the expected amount from payment hash
+        $payment_hash_data = $this->payment_hash->data;
+        $expected_amount = $payment_hash_data->amount_with_fee;
+
+        // ALWAYS adjust invoice allocations to match actual received amount
+        if ($fiat_amount != $expected_amount) {
```

Comment:
I've been burnt on these types of comparators in the past.

I've implemented a way to compare floats, you'll want to use something like

```php
!BcMath::equal($fiat_amount, $expected_amount)
```